proto-lens-protoc/src/Data/ProtoLens/Compiler/Generate.hs

```diff
@@ -259,80 +259,185 @@ generateEnumExports syntaxType e = [exportAll n, exportWith n aliases]
+        , alias "enumFromThenTo"
+            "Data.ProtoLens.Message.Enum.messageEnumFromThenTo"
         ]
 
-    -- proto3 enums always default to zero.
+    -- instance Data.Default.Class.Default Foo where
+    --   def = FirstEnumValue
     , instDecl [] ("Data.Default.Class.Default" `ihApp` [dataType])
-        [[match "def" [] $ "Prelude.toEnum" @@ litInt 0]]
+        [[match "def" [] defaultCon]]
+    -- instance Data.ProtoLens.FieldDefault Foo where
+    --   fieldDefault = FirstEnumValue
     , instDecl [] ("Data.ProtoLens.FieldDefault" `ihApp` [dataType])
-        [[match "fieldDefault" [] $ "Prelude.toEnum" @@ litInt 0]]
-    ]
-    ++
-    -- pattern Value0 :: Foo
-    -- pattern Value0 = Foo 0
+        [[match "fieldDefault" [] defaultCon]]
+    ] ++
+    -- pattern Enum2a :: FooEnum
+    -- pattern Enum2a = Enum2
     concat
-        [ [ patSynSig n dataType
-          , patSyn (pVar n)
-                $ pApp (unQual dataName) [pLitInt k]
+        [ [ patSynSig aliasName dataType
+          , patSyn (pVar aliasName) (pVar originalName)
           ]
-        | v <- enumValues info
-        , let n = enumValueName v
-        , let k = fromIntegral $ enumValueDescriptor v ^. number
+        | EnumValueInfo
+            { enumValueName = aliasName
+            , enumAliasOf = Just originalName
+            } <- enumValues info
         ]
+
   where
-    dataName = enumName info
+    EnumInfo { enumName = dataName, enumDescriptor = ed } = info
+    errorMessage = "toEnum: unknown value for enum " ++ unpack (ed ^. name)
+                      ++ ": "
+
+    errorMessageExpr = "Prelude.error"
+                          @@ ("Prelude.++" @@ stringExp errorMessage
+                              @@ ("Prelude.show" @@ "k__"))
+    alias funName implName = [match funName [] implName]
+
     dataType = tyCon $ unQual dataName
 
+    unrecognizedValueName, unrecognizedName :: Name
+    unrecognizedValueName = modifyIdent (++ "'UnrecognizedValue") dataName
+    unrecognizedName = modifyIdent (++ "'Unrecognized") dataName
+
+    unrecognizedValueType = tyCon $ unQual unrecognizedValueName
+
+    constructors :: [(Name, EnumValueDescriptorProto)]
+    constructors = List.sortBy (comparing ((^. number) . snd))
+                            [(n, d) | EnumValueInfo
+                                { enumValueName = n
+                                , enumValueDescriptor = d
+                                , enumAliasOf = Nothing
+                                } <- enumValues info
+                            ]
+    constructorNames = map fst constructors
+
+    defaultCon = con $ unQual $ head constructorNames
+
+    minBoundName = head constructorNames
+    maxBoundName = last constructorNames
+
+    constructorNumbers = map (second (fromIntegral . (^. number))) constructors
+
+    succPairs = zip constructorNames $ tail constructorNames
+    succDecl funName boundName thePairs =
```

Comment:
Sounds good. Is not erroring on `succ maxEnum` and instead returning `SomeEnum'UnrecognizedValue... (fromEnum maxEnum + 1`) the desired behavior here?

Do you also want me to redefine `pred` as `toEnum . (-1) . fromEnum`? Or should `pred minEnum` still result in an error?

Reply:
I think this would break things, eg for `FooEnum` as defined in proto3.proto:
```
  enum FooEnum {
    option allow_alias = true;
    Enum1 = 0;
    Enum2 = 3;
    Enum2a = 3;
  }
```
it seems like we want `succ Enum1 == Enum2`, not `succ Enum1 == Foo'FooEnum'Unrecognized (Foo'FooEnum'UnrecognizedValue 2)`